octopus-server/src/main/java/octopus/teamcity/server/generic/CreateDeploymentStep.java

```diff
@@ -0,0 +1,74 @@
+package octopus.teamcity.server.generic;
+
+import java.util.List;
+import java.util.Map;
+
+import com.intellij.openapi.util.text.StringUtil;
+import jetbrains.buildServer.serverSide.InvalidProperty;
+import octopus.teamcity.common.createdeployment.CreateDeploymentPropertyNames;
+import octopus.teamcity.common.createdeployment.CreateDeploymentUserData;
+import org.apache.commons.compress.utils.Lists;
+
+public class CreateDeploymentStep extends OctopusBuildStep {
+
+  private final CreateDeploymentPropertyNames KEYS = new CreateDeploymentPropertyNames();
+
+  public CreateDeploymentStep() {
+    super(
+        "create-deployment",
+        "Create deployment",
+        "editCreateDeploymentParameters.jsp",
+        "viewCreateDeploymentParameters.jsp");
+  }
+
+  @Override
+  public String describeParameters(Map<String, String> properties) {
+    final CreateDeploymentUserData userData = new CreateDeploymentUserData(properties);
+
+    return String.format(
+        "Project: %s\nEnvironments: %s\nRelease version: %s",
+        userData.getProjectNameOrId(),
+        String.join(", ", userData.getEnvironmentIdsOrNames()),
+        userData.getReleaseVersion());
+  }
+
+  @Override
+  public List<InvalidProperty> validateProperties(Map<String, String> properties) {
+    final List<InvalidProperty> failedProperties = Lists.newArrayList();
+    final CreateDeploymentUserData userData = new CreateDeploymentUserData(properties);
+
+    if (StringUtil.isEmpty(userData.getProjectNameOrId().trim())) {
```

Comment:
TC will auto-trim - also, the getProjectNameOrId will throw if the field is null/missing - so I'm not sure you'll get this property appearing.

Reply:
Deployment step refactored to use property map directly. Trim usage removed as appropriate